packages/theme-language-server-common/src/rename/providers/LiquidVariableRenameProvider.ts

```diff
@@ -0,0 +1,149 @@
+import { BaseRenameProvider } from '../BaseRenameProvider';
+import { DocumentManager } from '../../documents';
+import {
+  LiquidHtmlNode,
+  LiquidTagFor,
+  LiquidTagTablerow,
+  NamedTags,
+  NodeTypes,
+  Position,
+} from '@shopify/liquid-html-parser';
+import { Range } from 'vscode-languageserver';
+import {
+  PrepareRenameParams,
+  PrepareRenameResult,
+  RenameParams,
+  TextDocumentEdit,
+  TextEdit,
+  WorkspaceEdit,
+} from 'vscode-languageserver-protocol';
+import { visit } from '../../visitor';
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import { JSONNode } from '@shopify/theme-check-common';
+
+export class LiquidVariableRenameProvider implements BaseRenameProvider {
+  constructor(private documentManager: DocumentManager) {}
+
+  async prepare(
+    node: LiquidHtmlNode,
+    ancestors: LiquidHtmlNode[],
+    params: PrepareRenameParams,
+  ): Promise<null | PrepareRenameResult> {
+    const document = this.documentManager.get(params.textDocument.uri);
+    const textDocument = document?.textDocument;
+
+    if (!textDocument || !node || !ancestors) return null;
+    if (!supportedTags(node)) return null;
+
+    return {
+      range: Range.create(
+        textDocument.positionAt(node.position.start),
+        textDocument.positionAt(node.position.end),
+      ),
+      placeholder: variableName(node),
+    };
+  }
+
+  async rename(
+    node: LiquidHtmlNode,
+    ancestors: LiquidHtmlNode[],
+    params: RenameParams,
+  ): Promise<null | WorkspaceEdit> {
+    const document = this.documentManager.get(params.textDocument.uri);
+    const textDocument = document?.textDocument;
+
+    if (!textDocument || !node || !ancestors) return null;
+    if (document.ast instanceof Error) return null;
+    if (!supportedTags(node)) return null;
+
+    const oldName = variableName(node);
+    const scope = variableNameBlockScope(oldName, ancestors);
+    const replaceRange = textReplaceRange(oldName, textDocument, scope);
+
+    const ranges: Range[] = visit(document.ast, {
+      VariableLookup: replaceRange,
+      AssignMarkup: replaceRange,
+      ForMarkup: replaceRange,
+    });
+
+    const textDocumentEdit = TextDocumentEdit.create(
+      { uri: textDocument.uri, version: textDocument.version },
+      ranges.map((range) => TextEdit.replace(range, params.newName)),
+    );
+
+    return {
+      documentChanges: [textDocumentEdit],
+    };
+  }
+}
+
+function supportedTags(node: LiquidHtmlNode): boolean {
```

Comment:
You could change the return signature to 
```ts
function supportedTags(node: LiquidHtmlNode): node is AssignMarkup | VariableLookup | ForMarkup {
  // ...
}```

This type guard would make allow you for the other functions to accept a subset of nodes instead of all of them.